### ISSUE TITLE
Exception closing app using unmodified SuspendAndResumeService - Unable to remove the file to be replaced.

### ISSUE THREAD

**Issue by michael-hawker** [Microsoft/WindowsTemplateStudio#2794]
### Repro steps
Use [SuspendAndResume service](https://github.com/Microsoft/WindowsTemplateStudio/blob/dev/templates/Uwp/Features/SuspendAndResume/Services/SuspendAndResumeService.cs) from WTS as-is, no modifications or hooking into event.
Run app.
Close app.

(I've only seen this once after running app dozens of times.)

### Expected Behavior
App closes without exception.

### Actual Behavior

"Unable to remove the file to be replaced. (Exception from HRESULT: 0x80070497)"
>	MyProject.exe!MyProject.App.InitializeComponent.AnonymousMethod__16_0(object sender, Windows.UI.Xaml.UnhandledExceptionEventArgs e) Line 53	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	System.Private.CoreLib.dll!System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal.RoReportUnhandledError(System.Runtime.InteropServices.WindowsRuntime.IRestrictedErrorInfo error)	Unknown
 	System.Private.CoreLib.dll!System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal.ReportUnhandledError(System.Exception e)	Unknown
 	System.Runtime.WindowsRuntime.dll!System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()	Unknown
 	System.Runtime.WindowsRuntime.dll!System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeInContext(object thisObj)	Unknown
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown
 	System.Runtime.WindowsRuntime.dll!System.Threading.WinRTSynchronizationContextBase.Invoker.Invoke()	Unknown

StackTrace:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MyProject.Helpers.SettingsStorageExtensions.<SaveAsync>d__2`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MyProject.Services.SuspendAndResumeService.<SaveStateAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at MyProject.App.<App_EnteredBackground>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

## System

* **VS Version:** 15.9.2
* **WTS Wizard Version:** 2.5.18319.1
* **WTS Template Version:** 
* **Windows Build:** 17763.134

**Comment by michael-hawker** [comment]
Thanks @sibille, makes sense.  Though in the case of failure, does that mean the data wouldn't be saved?  It'd be bad to rely on this call, have it fail unknowingly, and have a user of your app lose your data.

Is this intermittent failure a platform bug/issue?

**Comment by sibille** [comment]
We should add a TODO comment so the developer can handle as appropriate when this fails. 
We're calling StorageFolder.CreateFileAsync(filename, CreationCollisionOption.ReplaceExisting), 
Docs say this can file due to bad FileName or UnauthorizedAccess, but your error seems to be a system error too.
https://docs.microsoft.com/en-us/uwp/api/windows.storage.storagefolder.createfileasync#Windows_Storage_StorageFolder_CreateFileAsync_System_String_Windows_Storage_CreationCollisionOption_

**Comment by michael-hawker** [comment]
Based on my observations, I believe it's a concurrency issue and signifies the file is already opened by the same or another process.

Maybe the service could add some thread safety to ensure that any save calls aren't done concurrently?

**Comment by sibille** [comment]
How about doing both, add thread safety and error handling in case this fails for another reason.

**Comment by crutkas** [comment]
A big thing is how we validate and reproduce this scenario so we can test against it for the solution

**Comment by dgomezc** [comment]
We have secured the Suspend and Resume service, but we have not been able to reproduce the error.

**Comment by dgomezc** [comment]
Following up with engineering team in parallel. We will open a new issue if we can do anything else.

**Comment by sibille** [comment]
Verified in dev-nightly:
Templates version: 0.20.19021.3
Wizard version: 0.20.19021.3
